download.go

```diff
@@ -0,0 +1,314 @@
+	resp, err := client.Do(req)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("problem downloading file for url %s", info.URL))
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return errors.Errorf("%s: could not download %s to path %s", resp.Status, info.URL, info.Path)
+	}
+
+	if err = WriteFile(resp.Body, info.Path); err != nil {
+		return err
+	}
+
+	if info.ArchiveOpts.ShouldExtract {
+		if err = doExtract(info); err != nil {
+			return errors.Wrapf(err, "problem extracting file %s to path %s", info.Path, info.ArchiveOpts.TargetPath)
+		}
+	}
+
+	return nil
+}
+
+func doExtract(info DownloadInfo) error {
+	var archiveHandler archiver.Archiver
+	switch info.ArchiveOpts.Format {
+	case ArchiveAuto:
+		unzipper := archiver.MatchingFormat(info.Path)
+		if unzipper == nil {
+			return errors.Errorf("could not detect archive format for %s", info.Path)
+		}
+		archiveHandler = unzipper
+	case ArchiveTarGz:
+		archiveHandler = archiver.TarGz
+	case ArchiveZip:
+		archiveHandler = archiver.Zip
+	default:
+		return errors.Errorf("unrecognized archive format %s", info.ArchiveOpts.Format)
+	}
+
+	if err := archiveHandler.Open(info.Path, info.ArchiveOpts.TargetPath); err != nil {
+		return errors.Wrapf(err, "problem extracting archive %s to %s", info.Path, info.ArchiveOpts.TargetPath)
+	}
+
+	return nil
+}
+
+// SetupDownloadMongoDBReleases performs necessary setup to download MongoDB with the given options.
+func SetupDownloadMongoDBReleases(ctx context.Context, cache *lru.Cache, opts MongoDBDownloadOptions) error {
+	if err := MakeEnclosingDirectories(opts.Path); err != nil {
+		return errors.Wrap(err, "problem creating enclosing directories")
+	}
+
+	feed, err := bond.GetArtifactsFeed(opts.Path)
+	if err != nil {
+		return errors.Wrap(err, "problem making artifacts feed")
+	}
+
+	catcher := grip.NewBasicCatcher()
+	urls, errs1 := feed.GetArchives(opts.Releases, opts.BuildOpts)
+	jobs, errs2 := createDownloadJobs(opts.Path, urls)
+
+	if err := setupDownloadJobsAsync(ctx, jobs, processDownloadJobs(ctx, addMongoDBFilesToCache(cache, opts.Path))); err != nil {
+		catcher.Add(errors.Wrap(err, "problem starting download jobs"))
+	}
+
+	if err := aggregateErrors(errs1, errs2); err != nil {
+		catcher.Add(errors.Wrap(err, "problem initializing download jobs"))
+	}
+
+	return catcher.Resolve()
+}
+
+func createDownloadJobs(path string, urls <-chan string) (<-chan amboy.Job, <-chan error) {
+	output := make(chan amboy.Job)
+	errOut := make(chan error)
+
+	go func() {
+		catcher := grip.NewCatcher()
+		for url := range urls {
+			j, err := recall.NewDownloadJob(url, path, true)
+			if err != nil {
+				catcher.Add(errors.Wrapf(err, "problem creating download job for %s", url))
+				continue
+			}
+
+			output <- j
+		}
+		close(output)
+		if catcher.HasErrors() {
+			errOut <- catcher.Resolve()
+		}
+		close(errOut)
+	}()
+
+	return output, errOut
+}
+
+func aggregateErrors(groups ...<-chan error) error {
```

Comment:
you can just pass catcher around, they're threadsafe.

Reply:
Changed to use catcher.